spirv_glsl.cpp

```diff
@@ -718,6 +719,54 @@ void CompilerGLSL::emit_header()
 			statement("#extension ", ext, " : enable");
 			statement("#endif");
 		}
+		else if (ext == "GL_KHR_shader_subgroup_vote")
+		{
+			declaredKHR_shader_subgroup_vote = true;
+			if (options.vulkan_semantics)
+				statement("#extension GL_KHR_shader_subgroup_vote : require");
+			else
+			{
+				statement("#if defined(GL_ARB_shader_group_vote)");
+				statement("#extension GL_ARB_shader_group_vote : require");
+				statement("#elif defined(GL_NV_gpu_shader5)");
+				statement("#extension GL_NV_gpu_shader5 : require");
+				statement("#elif defined(GL_AMD_gcn_shader)&&defined(GL_AMD_gpu_shader_int64)");
+				statement("#extension GL_AMD_gcn_shader : require");
+				statement("#extension GL_AMD_gpu_shader_int64 : enable"); //needed?
+				statement("#else");
+				statement("#error No extensions available for subgroupAllEqual()");
+				statement("#endif");
+			}
+		}
+		else if (ext == "GL_KHR_shader_subgroup_basic")
+		{
+			declaredKHR_shader_subgroup_basic = true;
+			if (options.vulkan_semantics)
+				statement("#extension GL_KHR_shader_subgroup_basic : require");
+			else
+			{
+				statement("#if defined(GL_NV_shader_thread_group)");
+
+				statement("#define GL_NV_shader_thread_group");
+
+				statement("#elif defined(GL_ARB_shader_ballot)");
+
+				statement("#extension GL_ARB_shader_ballot : require");
+				statement("#extension GL_AMD_gcn_shader : enable");
+				statement("#ifdef GL_AMD_gcn_shader");
+				statement("#define gl_SubgroupSize uint(gl_SIMDGroupSizeAMD)");
+				statement("#else defined(GL_ARB_shader_ballot)");
+				statement("#define gl_SubgroupSize gl_SubGroupSizeARB");
+				statement("#endif");
+				statement("#define gl_SubgroupInvocationID gl_SubGroupInvocationARB");
```

Comment:
@HansKristian-Work is that the right place to put these?

Also should we `#define` or should we declare a `const uint` ?

Furthermore @Crisspl anything starting with `gl_` in GLSL is reserved, so I guess we'll need to use a different prefix like SPIRV-Cross uses for its `gl_InstanceIndex` replacement.

Reply:
> If you emit a constant, you can call it SPIRV_Cross_gl_something, and change the name where builtin name is queried depending on vulkan_semantic. You can emit these constants in emit_resources(). You should be able to detect if the capability is enabled and only emit builtin constants based on that.

https://github.com/KhronosGroup/SPIRV-Cross/issues/1352#issuecomment-621796811

Reply:
SubgroupSize is not a constant necessarily, so `#define` overriding it is fine.

Reply:
we don't mean spec constant or a uniform, we could just declare it as a `const uint SugroupSize = ;`

Should we prefix with `SPIRV_Cross_gl` instead of `gl`?

Reply:
> anything starting with gl_ in GLSL is reserved, so I guess we'll need to use a different prefix

I thought it applies to GLSL only, without preprocessor?

> If you emit a constant, you can call it SPIRV_Cross_gl_something, and change the name where builtin name is queried depending on vulkan_semantic. 

Change the name?

Reply:
I guess you need to test some driver compilers in OpenGL if they let you feed them GLSL core profile shaders with `#define` which define things which start with `gl_` and then use them in the code

Reply:
@Crisspl did you check if Vendor OpenGL runtime compilers will compile GLSL with a `#define` where the symbol starts with `gl_` ?

Reply:
ok I checked on NVidia and Intel, `#define` with a `gl_` prefix (LOWER CASE) is fine, upper case is not... so we're in the clear

spec is also worded so.

Reply:
@HansKristian-Work you're fine with this `#define gl_` ?